mars_troughs/mcmc.py

```diff
@@ -0,0 +1,239 @@
+        sys.stderr.flush()
+
+        start = time.time()
+        
+        #Initialize sampler
+        self.sampler = emcee.EnsembleSampler(self.nwalkers, self.ndim, 
+                                             self.ln_likelihood,  
+                                             parameter_names=self.parameter_names)
+        #Run MCMC and track progress
+        self.sampler.reset()
+        #Iteratively compute autocorrelation time Tau
+        index=0
+        self.autocorr=np.zeros(10)
+        self.taus=np.zeros((10,self.ndim))
+        old_tau=np.inf
+        
+        #compute tau every n iterations
+        for sample in self.sampler.sample(self.initParams,iterations=self.maxSteps, 
+        progress=False):
+        
+            if self.sampler.iteration%(self.maxSteps/10):
+                continue
+            sys.stderr.flush()
+            print(self.sampler.iteration/self.maxSteps*100,'%',file=sys.stderr)
+            
+            tau=self.sampler.get_autocorr_time(tol=0)
+            self.autocorr[index]=np.mean(tau)
+            self.taus[index,:]=tau
+            index+=1
+            converged=np.all(tau*50<self.sampler.iteration)
+            converged&=np.all(np.abs(old_tau-tau)/tau<0.01)
+            
+            if converged:
+                break
+            
+            old_tau=tau
+            
+        end = time.time()
+        running_time=end-start
+
+        print("Running time {0:.1f} seconds".format(running_time), 
+              'for',self.modelName,self.maxSteps,file=sys.stderr)
+        sys.stderr.flush()
+        
+        
+        #save samples and log prob every thin_by steps
+        all_samples=self.sampler.get_chain()
+        self.samples=all_samples[0::self.thin_by,:,:]
+        all_logprob=self.sampler.get_log_prob()
+        self.logprob=all_logprob[0::self.thin_by,:]
+        self.accFraction=self.sampler.acceptance_fraction
+        self.totalSteps=self.sampler.iteration
+        
+        #delete sampler because it is very large
+        del self.sampler
+        
+
+    def ln_likelihood(self,params: Dict[str,float]):
+        
+        errorbar: float = params["errorbar"]
+        
+        if errorbar < 0: #prior on the variance (i.e. the error bars)
+            return -1e99
+        
+        #set trough model with candidate parameters
+        self.tr.set_model(params)
+        #compute likelihood of model 
+        likelihood_of_model=self.tr.lnlikelihood(self.xdata,self.ydata,
+                                                 self.tr.accuModel._times)
+        #prior lag with time
+        if any(self.tr.lag_at_t  < 1e-15) or any(self.tr.lag_at_t > 20):
+            return -1e99
+        
+        #prior nearest points to observed data
+        if any(self.tr.ynear < -2e3) or any(self.tr.ynear > 0):
+            return -1e99
+        
+        #get accumulation with time
+        acc_t=self.tr.accuModel.get_accumulation_at_t(self.tr.accuModel._times)
```

Comment:
If this call is contributing to the slow down then we can cache these values on the trough object and fetch them, rather than recomputing them. Does that make sense?